File: LowLevelDesign/java/task-scheduler/src/com/taskscheduler/observer/TaskExecutionObserver.java
package com.taskscheduler.observer;

import com.taskscheduler.ScheduledTask;

public interface TaskExecutionObserver {
    void onTaskStarted(ScheduledTask task);
    void onTaskCompleted(ScheduledTask task);
    void onTaskFailed(ScheduledTask task, Exception e);
}

File: LowLevelDesign/task-scheduler-nodejs/src/observer/TaskExecutionObserver.js
/**
 * TaskExecutionObserver interface - defines callbacks for task lifecycle events
 */
export class TaskExecutionObserver {
    onTaskStarted(task) {
        throw new Error("onTaskStarted() must be implemented by subclass");
    }

    onTaskCompleted(task) {
        throw new Error("onTaskCompleted() must be implemented by subclass");
    }

    onTaskFailed(task, error) {
        throw new Error("onTaskFailed() must be implemented by subclass");
    }
}
